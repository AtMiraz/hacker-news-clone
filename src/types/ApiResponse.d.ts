export type Hit = {
  created_at: string
  title: string | null
  url: string | null
  author: string
  points: string | null
  story_text: string | null
  comment_text: string | null
  num_comments: number | null
  story_id: number
  story_title: string
  story_url: string
  parent_id: number
  created_at_i: number
  _tags: string[]
  objectID: string
  _highlightResult: any
}
