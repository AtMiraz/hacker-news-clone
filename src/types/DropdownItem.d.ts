import { News } from './StoryList'

export type dropdownItem = {
  name: string
  value: News
  image: string
  selected: boolean
}
