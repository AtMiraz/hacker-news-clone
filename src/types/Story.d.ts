export type Story = {
  storyId: number
  title: string | null
  author: string
  createdAt: string
  url: string | null
  favorite: boolean
}
