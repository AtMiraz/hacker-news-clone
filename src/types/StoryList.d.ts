export type News = 'angular' | 'reactjs' | 'vuejs'

export type StoryListProps = {
  mode?: 'all' | 'favorites'
  newsType?: News
}
