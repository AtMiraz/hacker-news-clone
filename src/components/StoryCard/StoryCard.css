.hacker__news__app-story-card {
  width: 34rem;
  height: 6.262rem;
  border: solid 1px #979797;
  border-radius: 6px;
  background-color: #fff;
  display: flex;
  justify-content: space-between;
  position: relative;
  min-width: 480px;
}

.story__card__info {
  cursor: pointer;
  width: 30.125rem;
}
.story__card__info-author {
  margin-left: 1.625rem;
  margin-top: 1.625rem;
  font-family: Roboto;
  font-size: 0.688rem;
  color: #767676;
  display: flex;
  flex-direction: row;
}

.story__card__info-author > img {
  margin-right: 0.5rem;
}

.story__card__info-title {
  margin-left: 1.625rem;
  font-family: Roboto;
  font-size: 0.875rem;
  font-weight: 500;
  letter-spacing: 0.25px;
  color: #6b6b6b;
  margin-top: 0.438rem;
  max-width: 25rem;
  text-overflow: ellipsis;
  overflow: hidden;
  word-break: break-all;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  display: -webkit-box;
}

.story__card__actions {
  width: 4.35rem;
  height: 100%;
  margin-left: auto;
  margin-right: 3px;
  opacity: 0.8;
  border-left: solid 1px #979797;
  background-color: #fcfcfc;
  display: flex;
  justify-content: center;
  align-items: center;
}

.story__card__actions > img {
  width: 1.5rem;
  height: 1.375rem;
  cursor: pointer;
}

.story__card__actions-heart:not(:active) {
  transition: all 0.5s ease-in-out;
  transform: scale(1.1);
}
