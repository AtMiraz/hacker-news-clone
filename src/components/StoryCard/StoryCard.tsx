import './StoryCard.css'
import clock from '../../assets/svgs/clock.svg'
import heartFilled from '../../assets/svgs/heart-filled.svg'
import heartOutline from '../../assets/svgs/heart-outline.svg'
import { Story } from '../../types/Story'
import { formatDistance } from 'date-fns'

type StoryProps = {
  story: Story
  onClick: (story: Story) => void
}

function openInNewWindow(story: Story): void {
  const url =
    story.url != null
      ? story.url
      : `https://news.ycombinator.com/item?id=${story.storyId}`

  // mitigates security risk of opening url on _blank
  // https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/
  const newWindow = window.open(url, '_blank', 'noopener,noreferrer')
  if (newWindow) newWindow.opener = null
}

function handleDate(dateString: string): string {
  const date = new Date(dateString)
  return formatDistance(new Date(), date, { addSuffix: false })
}

function StoryCard({ story, onClick }: StoryProps) {
  return (
    <div
      className="hacker__news__app-story-card"
      data-testid="story-card-component"
    >
      <div className="story__card__info" onClick={() => openInNewWindow(story)}>
        <div className="story__card__info-author">
          <img src={clock} alt="clock" />
          <span data-testid="story-info">
            {handleDate(story.createdAt)} ago by {story.author}
          </span>
        </div>
        <div className="story__card__info-title" data-testid="story-title">
          {story.title}
        </div>
      </div>
      <div className="story__card__actions">
        <img
          className="story__card__actions-heart"
          data-testid="heart-button"
          onClick={() => onClick(story)}
          src={story.favorite ? heartFilled : heartOutline}
          alt="favorited"
        />
      </div>
    </div>
  )
}

export default StoryCard
