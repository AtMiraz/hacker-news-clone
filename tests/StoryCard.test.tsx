import React from 'react'
import StoryCard from '../src/components/StoryCard/StoryCard'
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/react'
import { expect, test, describe, vi } from 'vitest'
import { Story } from '../src/types/Story'

describe('StoryCard tests', () => {
  const story: Story = {
    title: 'some title',
    author: 'an author',
    createdAt: '2020-05-15T20:57:04.000Z',
    favorite: false,
    url: 'some url',
  }

  test('Should render the component provided the right data', () => {
    const component = render(
      <StoryCard
        story={story}
        onClick={() => {
          return
        }}
      />
    )
    expect(component).toMatchSnapshot()
    expect(component).to.exist
  })

  test('Should trigger a function when user clicks on the heart', async () => {
    const onClick = vi.fn()
    render(<StoryCard story={story} onClick={onClick} />)

    const heart = await screen.findByTestId('heart-button')
    await userEvent.click(heart)

    expect(onClick).toHaveBeenCalledOnce
  })

  test('should display the correct information on the card', async () => {
    render(
      <StoryCard
        story={story}
        onClick={() => {
          return
        }}
      />
    )

    const title = await screen.findByTestId('story-title')
    const info = await screen.findByTestId('story-info')

    expect(title.textContent).toContain('some title')
    expect(info.textContent).toContain('about 3 years ago by an author')
  })
})
