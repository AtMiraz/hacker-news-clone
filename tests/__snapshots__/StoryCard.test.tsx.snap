// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StoryCard tests > Should render the component provided the right data 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hacker__news__app-story-card"
        data-testid="story-card-component"
      >
        <div
          class="story__card__info"
        >
          <div
            class="story__card__info-author"
          >
            <img
              alt="clock"
              src="/src/assets/svgs/clock.svg"
            />
            <span
              data-testid="story-info"
            >
              about 3 years
               ago by 
              an author
            </span>
          </div>
          <div
            class="story__card__info-title"
            data-testid="story-title"
          >
            some title
          </div>
        </div>
        <div
          class="story__card__actions"
        >
          <img
            alt="favorited"
            class="story__card__actions-heart"
            data-testid="heart-button"
            src="/src/assets/svgs/heart-outline.svg"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="hacker__news__app-story-card"
      data-testid="story-card-component"
    >
      <div
        class="story__card__info"
      >
        <div
          class="story__card__info-author"
        >
          <img
            alt="clock"
            src="/src/assets/svgs/clock.svg"
          />
          <span
            data-testid="story-info"
          >
            about 3 years
             ago by 
            an author
          </span>
        </div>
        <div
          class="story__card__info-title"
          data-testid="story-title"
        >
          some title
        </div>
      </div>
      <div
        class="story__card__actions"
      >
        <img
          alt="favorited"
          class="story__card__actions-heart"
          data-testid="heart-button"
          src="/src/assets/svgs/heart-outline.svg"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
